### ISSUE TITLE
OutOfMemoryError has been occurred

### ISSUE THREAD

**Issue by hehexiansheng123** [eclipse-ee4j/angus-mail#158]
My product test is connected to a self-built email server. When verifying the email account or sending emails, a large packet is returned, causing memory overflow. Does this software verify the packet length in these two scenarios? If it is not verified, it is vulnerable to attacks.

Thanks

**Comment by hehexiansheng123** [comment]
According to the code, the response from the email server is not verified.

**Comment by hehexiansheng123** [comment]
According to the code, it seems that the response from the email server is not verified.

protected int readServerResponse() throws MessagingException {
	assert Thread.holdsLock(this);
        String serverResponse = "";
        int returnCode = 0;
	StringBuilder buf = new StringBuilder(100);

	// read the server response line(s) and add them to the buffer
	// that stores the response
        try {
	    String line = null;

	    do {
		line = lineInputStream.readLine();
		if (line == null) {
		    serverResponse = buf.toString();
		    if (serverResponse.length() == 0)
			serverResponse = "[EOF]";
		    lastServerResponse = serverResponse;
		    lastReturnCode = -1;
		    logger.log(Level.FINE, "EOF: {0}", serverResponse);
		    return -1;
		}
		buf.append(line);
		buf.append("\n");
	    } while (isNotLastLine(line));

            serverResponse = buf.toString();
        } catch (IOException ioex) {
	    logger.log(Level.FINE, "exception reading response", ioex);
            //ioex.printStackTrace(out);
	    lastServerResponse = "";
	    lastReturnCode = 0;
	    throw new MessagingException("Exception reading response", ioex);
            //returnCode = -1;
        }

	// print debug info
        //if (logger.isLoggable(Level.FINE))
            //logger.fine("RCVD: " + serverResponse);

	// parse out the return code
        if (serverResponse.length() >= 3) {
            try {
                returnCode = Integer.parseInt(serverResponse.substring(0, 3));
            } catch (NumberFormatException nfe) {
		try {
		    close();
		} catch (MessagingException mex) {
		    // thrown by close()--ignore, will close() later anyway
		    logger.log(Level.FINE, "close failed", mex);
		}
		returnCode = -1;
            } catch (StringIndexOutOfBoundsException ex) {
		try {
		    close();
		} catch (MessagingException mex) {
		    // thrown by close()--ignore, will close() later anyway
		    logger.log(Level.FINE, "close failed", mex);
		}
                returnCode = -1;
	    }
	} else {
	    returnCode = -1;
	}
	if (returnCode == -1)
	    logger.log(Level.FINE, "bad server response: {0}", serverResponse);

        lastServerResponse = serverResponse;
	lastReturnCode = returnCode;
        return returnCode;
    }

**Comment by jbescos** [comment]
@hehexiansheng123 could you attach the stacktrace of the OutOfMemoryError, please?. I know that could happen in an unrelated area, but just in case it happens in the place we need to take a look...

**Comment by baoxinggit** [comment]
Hello, sorry to reply now,
this is the stack information:
![堆栈信息](https://github.com/user-attachments/assets/3de8bdb8-5795-44a8-a6b1-e36d68275552)
overflow area:
![堆信息](https://github.com/user-attachments/assets/0e7223e7-0ac5-4cb4-b4ce-661ef8276c58)

**Comment by hehexiansheng123** [comment]
> Hello, sorry to reply now, this is the stack information: ![堆栈信息](https://private-user-images.githubusercontent.com/26064820/381991444-3de8bdb8-5795-44a8-a6b1-e36d68275552.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0MjU3MDgsIm5iZiI6MTczMDQyNTQwOCwicGF0aCI6Ii8yNjA2NDgyMC8zODE5OTE0NDQtM2RlOGJkYjgtNTc5NS00NGE4LWE2YjEtZTM2ZDY4Mjc1NTUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAxVDAxNDMyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkMGI5MjVmYjI0NjgzMDlkMjVmOWE0ZjZlMjkwMWZhMjU4OWVmNWNjZTljMjg0MzI5ZjIzNmM3NzkwNmI4MDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VVx7liLZPNCz7DpRrUF3glf5uY8Ak58iN4BVmUMQWgo) overflow area: ![堆信息](https://private-user-images.githubusercontent.com/26064820/381994349-0e7223e7-0ac5-4cb4-b4ce-661ef8276c58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0MjU3MDgsIm5iZiI6MTczMDQyNTQwOCwicGF0aCI6Ii8yNjA2NDgyMC8zODE5OTQzNDktMGU3MjIzZTctMGFjNS00Y2I0LWI0Y2UtNjYxZWY4Mjc2YzU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAxVDAxNDMyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4ZTM3MWY1NzdlYWJhNjUyMjVmZDE2OWQ4ODdlYzI1ZjkyMTI4YTE0NDQxN2U0ODE2Y2IwOTFmZDRiZGNiYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gMMfY0ZFmTCuhrGKGcD9FNNzZlLkCno_Gec4fKxNYTE)

test

**Comment by hehexiansheng123** [comment]
> Hello, sorry to reply now, this is the stack information: ![堆栈信息](https://private-user-images.githubusercontent.com/26064820/381991444-3de8bdb8-5795-44a8-a6b1-e36d68275552.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0MjU3MDgsIm5iZiI6MTczMDQyNTQwOCwicGF0aCI6Ii8yNjA2NDgyMC8zODE5OTE0NDQtM2RlOGJkYjgtNTc5NS00NGE4LWE2YjEtZTM2ZDY4Mjc1NTUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAxVDAxNDMyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkMGI5MjVmYjI0NjgzMDlkMjVmOWE0ZjZlMjkwMWZhMjU4OWVmNWNjZTljMjg0MzI5ZjIzNmM3NzkwNmI4MDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VVx7liLZPNCz7DpRrUF3glf5uY8Ak58iN4BVmUMQWgo) overflow area: ![堆信息](https://private-user-images.githubusercontent.com/26064820/381994349-0e7223e7-0ac5-4cb4-b4ce-661ef8276c58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0MjU3MDgsIm5iZiI6MTczMDQyNTQwOCwicGF0aCI6Ii8yNjA2NDgyMC8zODE5OTQzNDktMGU3MjIzZTctMGFjNS00Y2I0LWI0Y2UtNjYxZWY4Mjc2YzU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAxVDAxNDMyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4ZTM3MWY1NzdlYWJhNjUyMjVmZDE2OWQ4ODdlYzI1ZjkyMTI4YTE0NDQxN2U0ODE2Y2IwOTFmZDRiZGNiYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gMMfY0ZFmTCuhrGKGcD9FNNzZlLkCno_Gec4fKxNYTE)

@jbescos my colleagues had upload the stack information, take a look please, thanks bro

**Comment by jbescos** [comment]
Thank you both, @hehexiansheng123 and @baoxinggit 

@jmehrens , I am thinking to provide a new property with a default value in `LineInputStream` to limit the bytes to be read.

Do you have any suggestion?.

**Comment by hehexiansheng123** [comment]
> Thank you both, @hehexiansheng123 and @baoxinggit
> 
> @jmehrens , I am thinking to provide a new property with a default value in `LineInputStream` to limit the bytes to be read.
> 
> Do you have any suggestion?.

it seems a good idea, I think it's best to add a read length limit to the response body in the readLine method.

**Comment by jmehrens** [comment]
This thread talks a bit on limits:
https://github.com/jstedfast/MailKit/issues/834

**Comment by baoxinggit** [comment]
Sorry, what is the plan for solving this problem?

**Comment by jmehrens** [comment]
@baoxinggit 
> Sorry, what is the plan for solving this problem?

Looks like there was a PR submitted for this issue.  I linked it to this issue.
MailKit uses a 16k limit before failing and cites a RFC stating that there is a recommend limit of 1K and 8K for server responses but I wasn't able to locate the the RFC being cited.

I would think adding a new property with 16K limit default would be something to test out as I'm not sure what other details I'm missing on this change.

**Comment by baoxinggit** [comment]
May I ask when this issue can be merged and a new version released?

**Comment by jmehrens** [comment]
> May I ask when this issue can be merged and a new version released?

Feedback was given in the PR no new changes were pushed.  That is current blocker.

**Comment by baoxinggit** [comment]
Hi @jmehrens, just circling back on this issue as it's impacting our production environment. 
Could you provide a rough timeline?
We understand maintainers have many priorities - any guidance would be appreciated.

**Comment by jmehrens** [comment]
@baoxinggit the linked PR is waiting on changes from @tbelaire who created the PR.  Looks like the PR is abandoned.

This could move forward under a new PR with someone willing to work on it.  I can review and shepherd but won't have time to author a new PR.

**Comment by baoxinggit** [comment]
I found that using extension points can solve this problem. I used mail.smtp.socketFactory to extend the socket and checked the traffic at the getInputStream position. If the traffic exceeds the threshold, the connection is disconnected. Thank you for your feedback.
